x/dispensation/handler_test.go

```diff
@@ -74,6 +75,196 @@ func TestNewHandler_CreateDistribution_MultipleTypes(t *testing.T) {
+
+	msgLiquidityMining := types.NewMsgCreateDistribution(distributor, types.DistributionType_DISTRIBUTION_TYPE_LIQUIDITY_MINING, outputList, "")
+	res, err = handler(ctx, &msgLiquidityMining)
+	require.NoError(t, err)
+	require.NotNil(t, res)
+	res, err = handler(ctx, &msgLiquidityMining)
+	require.Error(t, err)
+	require.Nil(t, res)
+	msgLiquidityMiningOutput := msgLiquidityMining.Output
+	assert.Equal(t, recipients, len(msgLiquidityMiningOutput))
+	for i := 0; i < len(msgLiquidityMiningOutput); i++ {
+		assert.True(t, msgLiquidityMiningOutput[i].Coins.AmountOf("catk").Equal(sdk.NewInt(10)) ||
+			msgLiquidityMiningOutput[i].Coins.AmountOf("ceth").Equal(sdk.NewInt(10)))
+	}
+
+	msgLiquidityMining = types.NewMsgCreateDistribution(distributor, types.DistributionType_DISTRIBUTION_TYPE_LIQUIDITY_MINING, outputList, "")
+	_, err = handler(ctx, &msgLiquidityMining)
+	assert.Error(t, err, "Failed in collecting funds")
+
+	err = sifapp.AddCoinsToAccount(types.ModuleName, app.BankKeeper, ctx, distributor, totalCoins)
+	assert.NoError(t, err)
+
+	msgUnspecified := types.NewMsgCreateDistribution(distributor, types.DistributionType_DISTRIBUTION_TYPE_UNSPECIFIED, outputList, "")
+	res, err = handler(ctx, &msgUnspecified)
+	require.NoError(t, err)
+	require.NotNil(t, res)
+	res, err = handler(ctx, &msgUnspecified)
+	require.Error(t, err)
+	require.Nil(t, res)
+	msgUnspecifiedOutput := msgUnspecified.Output
+	assert.Equal(t, recipients, len(msgUnspecifiedOutput))
+	for i := 0; i < len(msgUnspecifiedOutput); i++ {
+		assert.True(t, msgUnspecifiedOutput[i].Coins.AmountOf("catk").Equal(sdk.NewInt(10)) ||
+			msgUnspecifiedOutput[i].Coins.AmountOf("ceth").Equal(sdk.NewInt(10)))
+	}
+
+	msgUnspecified = types.NewMsgCreateDistribution(distributor, types.DistributionType_DISTRIBUTION_TYPE_UNSPECIFIED, outputList, "")
+	_, err = handler(ctx, &msgUnspecified)
+	assert.Error(t, err, "Failed in collecting funds")
+
+	err = sifapp.AddCoinsToAccount(types.ModuleName, app.BankKeeper, ctx, distributor, totalCoins)
+	assert.NoError(t, err)
+
+	msgValidatorSubsidy := types.NewMsgCreateDistribution(distributor, types.DistributionType_DISTRIBUTION_TYPE_VALIDATOR_SUBSIDY, outputList, "")
+	res, err = handler(ctx, &msgValidatorSubsidy)
+	require.NoError(t, err)
+	require.NotNil(t, res)
+	res, err = handler(ctx, &msgValidatorSubsidy)
+	require.Error(t, err)
+	require.Nil(t, res)
+	msgValidatorSubsidyOutput := msgValidatorSubsidy.Output
+	assert.Equal(t, recipients, len(msgValidatorSubsidyOutput))
+	for i := 0; i < len(msgValidatorSubsidyOutput); i++ {
+		assert.True(t, msgValidatorSubsidyOutput[i].Coins.AmountOf("catk").Equal(sdk.NewInt(10)) ||
+			msgValidatorSubsidyOutput[i].Coins.AmountOf("ceth").Equal(sdk.NewInt(10)))
+	}
+
+	msgValidatorSubsidy = types.NewMsgCreateDistribution(distributor, types.DistributionType_DISTRIBUTION_TYPE_VALIDATOR_SUBSIDY, outputList, "")
+	_, err = handler(ctx, &msgValidatorSubsidy)
+	assert.Error(t, err, "Failed in collecting funds")
+}
```

Comment:
Can this be rewritten as a test table with several cases?

Reply:
So I tried to use test table here. But it is pretty hard to do that for this case, because we use randomize logic when try to distribute coins to users, so the output is not fixed.